core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala

```diff
@@ -720,7 +717,24 @@ object SparkSubmit extends CommandLineUtils {
       printWarning("Subclasses of scala.App may not work correctly. Use a main() method instead.")
     }
 
-    val mainMethod = mainClass.getMethod("main", new Array[String](0).getClass)
+    val sparkAppMainMethodArr = mainClass.getMethods().filter(_.getName == "sparkMain")
```

Comment:
Hmmm... instead of this, how about:

```
val sparkAppMain = mainClass.getMethods().find(_.getName == "sparkMain")
```

Then you have an `Option[Method]` which should make the code below a little cleaner. At the very least you won't need the `isSparkApp` variable nor the array dereference in L734. L733 would just become `mainMethod = sparkAppMain.getOrElse(...)`.